csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -0,0 +1,149 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using IceInternal;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Ice
+{
+    using Context = Dictionary<string, string>;
+
+    /// <summary>Represents a request protocol frame sent by the application.</summary>
+    public sealed class OutgoingRequestFrame : OutputStream
+    {
+        /// <summary>The identity of the target Ice object.</summary>
+        public Identity Identity { get; }
+
+        /// <summary>The facet of the target Ice object.</summary>
+        public string Facet { get; }
+
+        /// <summary>The operation called on the Ice object.</summary>
+        public string Operation { get; }
+
+         /// <summary>When true, the operation is idempotent.</summary>
+        public bool IsIdempotent { get; }
+
+        /// <summary>The request context. Its initial value is computed when the request frame is created.</summary>
+        public Context Context { get; }
+
+        /// <summary>An OutputStream is sealed when it can no longer be written into, in particular the payload
+        /// is complete and the context is written. Modidying Context once IsSealed is true has no effect.</summary>
+        public bool IsSealed { get; private set; } = false; // TODO: move to OutputStream with protected set
+
+        /// <summary>Creates a new outgoing request frame with an empty payload for the in-parameters.</summary>
+        /// <param name="proxy">A proxy to the target Ice object. This method uses the communicator, identity, facet,
+        /// encoding and context of this proxy to create the request frame.</param>
+        /// <param name="operation">The operation to invoke on the target Ice object.</param>
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame with an empty payload.</returns>
```

Comment:
did you consider using a default value of `false` for idempotent

Reply:
I prefer to keep idempotent explicit, especially with the latest ctors.